OSD prepare job cannot create vols with /dev/mapper alias on PVCs

ausias-armesto opened the pull request:
<!-- Please take a look at our [Contributing](https://rook.io/docs/rook/latest/Contributing/development-flow/)
documentation before submitting a Pull Request!
Thank you for contributing to Rook! -->

**Description of your changes:**
The kubernetes job `rook-ceph-osd-prepare-<server-name>` is not able to complete correctly when trying to create a OSD whose device is created by a LVM and has an alias starting by `/dev/mapper`. The job fails trying to create a `ceph-volume`

**Which issue is resolved by this Pull Request:**
Resolves #11297

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure)).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

parth-gr commented:
You also need to remove this check from the unit test https://github.com/rook/rook/blob/master/pkg/daemon/ceph/osd/volume_test.go
So the unit-test CI should pass

BlaineEXE commented:
If I understand correctly, this only affects PVC-based OSDs. Is that correct and intended? If so, I would request the title and commit subject be sure to clarify that it is only the PVC OSDs.

BlaineEXE commented:
I also think @satoru-takeuchi 's input will be invaluable here. As long as this is limited to PVC-based OSDs, I think this is safe, but I'm not sure if there are side-effects I'm not foreseeing.

satoru-takeuchi commented:
@ausias-armesto Does this PR ready to be reviewed?

In addition, please reflect [this comment](https://github.com/rook/rook/pull/11301#issuecomment-1315623377) if necessary.

BlaineEXE commented:
I see this commit body:

> Remove the condition that select the devices name prefixed with /dev/mapper as it is preventing the ceph-volume command to fail.

The way I read this, I think `preventing` is intended to be `causing`> is that right?

ausias-armesto commented:
> If I understand correctly, this only affects PVC-based OSDs. Is that correct and intended? If so, I would request the title and commit subject be sure to clarify that it is only the PVC OSDs.

Yes, that is the case that I am facing, for PVC based OSD. I've updated the commit message and fixed the unit test so you can review again

ausias-armesto commented:
Hi @satoru-takeuchi @BlaineEXE ,
Is there anything else that I can do to get approved this PR

satoru-takeuchi commented:
@ausias-armesto I've not reviewed this yet. Please wait for a while. I'll look into it in the next week.

satoru-takeuchi commented:
@ausias-armesto Please wait for several days to complete the review. The OSD creation code is complicated very much and I'm afraid regressions.

satoru-takeuchi commented:
I assume the root cause is not the treatment of "/dev/mapper" but the improper handling of `osdsPerDevice` field. Although OSD on LV doesn't support `osdsPerDevice` field (it's not documented yet and I'll fix it later), users can specify this field and Rook tries to osd by mistake.

I guess you can create an OSD without applying this PR with the following configuration which you mentioned. Could you confirm this?

```
          storage:
            useAllNodes: false
            useAllDevices: false
            # deviceFilter:
            nodes:
              - name: "server-stage-fsn1-002"
                devices:
                  - name: "/dev/vg0/data"
                    config:
                      osdsPerDevice: "1"
                      deviceClas: "hdd"
```

And the following configuration results in the problem that you saw. The difference is the number of `osdsPerDevice`.

```
          storage:
            useAllNodes: false
            useAllDevices: false
            # deviceFilter:
            nodes:
              - name: "server-stage-fsn1-002"
                devices:
                  - name: "/dev/vg0/data"
                    config:
                      osdsPerDevice: "3"
                      deviceClas: "hdd"
```

This can be verified by reading your log of the prepare pod.

```
2022-11-10 22:04:17.199022 D | cephosd: won't use raw mode since osd per device is 3
```

I guess that your problem will disappear if `osdsPerDevice` is not specified or is set to 1. Please try this.

satoru-takeuchi commented:
@travisn 

> Shouldn't we remove the dev mappings handling code for OSDs on PVCs? Or why would we need it?

You're right. This logic is not necessary. So this patch makes sense not as a bug fix but as a cleanup. How about merging this PR anyway? In addition, I'll submit a PR to handle `osdsPerDevice` properly.

satoru-takeuchi commented on pkg/daemon/ceph/osd/volume.go:
```diff
@@ -586,12 +579,6 @@ func (a *OsdAgent) initializeDevicesLVMMode(context *clusterd.Context, devices *
 			deviceArg := path.Join("/dev", name)
-			// ceph-volume prefers to use /dev/mapper/<name> if the device has this kind of alias
-			for _, devlink := range device.PersistentDevicePaths {
-				if strings.HasPrefix(devlink, "/dev/mapper") {
```
@travisn I only focused on LV-backed PVC by mistake and now I have a concern.

This logic is not necessary for LV-backed OSD for both OSD on device 
nd OSD on PVC because this configuration always uses raw mode and `ceph-volume raw prepare` accepts all device names. However, if a new OSD creation uses lvm mode, /dev/mapper prefix must be used because `ceph-volume lvm prepare` only accept the device names prefixed by /dev/mapper. This limitation exists for both OSD on device and OSD on PCV.

Please wait for a while. I'll verify my concern later.

ausias-armesto replied:
@travisn , I can only test on LV-backed PVC, and cannot provide further information about how would behave in other contexts. On the other side, changing `osdsPerDevice ` to `1` has not changed anything, is still not working.

Thanks for following this issue

satoru-takeuchi replied:
@ausias-armesto Could you show me your cluster.yaml and the log of osd prepare pod when you specified `osdsPerDevice: 1`?

ausias-armesto replied:
[osd-prepare-1.log](https://github.com/rook/rook/files/10192870/osd-prepare-1.log)
[osd-prepare-2.log](https://github.com/rook/rook/files/10192871/osd-prepare-2.log)
[osd-prepare-3.log](https://github.com/rook/rook/files/10192873/osd-prepare-3.log)
[helm-cluster.yaml.txt](https://github.com/rook/rook/files/10192942/helm-cluster.yaml.txt)

The disk partitions for servers running on osd2 and osd3 looks like this:
```
root@server-stage-fsn1-002 /home/debian # lsblk -f
NAME           FSTYPE            FSVER    LABEL    UUID                                   FSAVAIL FSUSE% MOUNTPOINT
sda                                                                                                      
├─sda1         linux_raid_member 1.2      rescue:0 492095b8-f9ee-e7c0-9e1e-e19193d3c00f                  
│ └─md0        ext3              1.0               ce3f530e-9eb8-4d67-ac74-44fe208f9328    877.3M     6% /boot
└─sda2         linux_raid_member 1.2      rescue:1 d9d405cd-528f-6f96-99b3-9b7fb7469523                  
  └─md1        LVM2_member       LVM2 001          ECx2sJ-9et7-Sg4F-oC59-nge2-LLDQ-3WACPR                
    ├─vg0-root ext4              1.0               0545917f-d52f-4902-81f6-88c5f02e84ba     37.9G    17% /
    ├─vg0-swap swap              1                 236c4901-9f82-4399-a8d6-b61147818169                  
    ├─vg0-tmp  reiserfs          3.6               9b4771e7-73a5-41f9-ab4f-56972f9be3e1      4.6G     8% /tmp
    └─vg0-data                                                                                           
sdb                                                                                                      
├─sdb1         linux_raid_member 1.2      rescue:0 492095b8-f9ee-e7c0-9e1e-e19193d3c00f                  
│ └─md0        ext3              1.0               ce3f530e-9eb8-4d67-ac74-44fe208f9328    877.3M     6% /boot
└─sdb2         linux_raid_member 1.2      rescue:1 d9d405cd-528f-6f96-99b3-9b7fb7469523                  
  └─md1        LVM2_member       LVM2 001          ECx2sJ-9et7-Sg4F-oC59-nge2-LLDQ-3WACPR                
    ├─vg0-root ext4              1.0               0545917f-d52f-4902-81f6-88c5f02e84ba     37.9G    17% /
    ├─vg0-swap swap              1                 236c4901-9f82-4399-a8d6-b61147818169                  
    ├─vg0-tmp  reiserfs          3.6               9b4771e7-73a5-41f9-ab4f-56972f9be3e1      4.6G     8% /tmp
    └─vg0-data                                                                                           
```


The server running on osd1, has 2 extra disks `sdc` and `sdd` which are correctly detected.

satoru-takeuchi commented:
@ausias-armesto Commitlint claims "body must have leading blank line [body-leading-blank]", please check it.

https://github.com/rook/rook/actions/runs/3693055432/jobs/6252601418#step:4:20

Other test failures seem to be intermittent.

satoru-takeuchi commented:
> Although OSD on LV doesn't support osdsPerDevice field (it's not documented yet and I'll fix it later), users can specify this field and Rook tries to osd by mistake.

Thi problem will be handled by #11438
